### PR TITLE
Add typing notifications to /sync responses - fixes #635

### LINKED ISSUE #574: [syncserver] Support typing events

### LINKED ISSUE #635: Make the syncapi consume messages from the typing server’s topic
Currently, while the typingserver seems to work correctly, the events it produces aren't consumed by the syncapi, which should consume them and include them in `/sync` responses.

### PR DESCRIPTION
**Pull request by Cnly**
This PR adds a new consumer for typing notifications in syncapi. It also brings changes to syncserver.go and some related files so EDUs can better fit in /sync responses.

Fixes #635.
Fixes #574.

Signed-off-by: Alex Chen <minecnly@gmail.com>

### Pull Request Checklist

* [x] I have added any new tests that need to pass to `testfile` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/CONTRIBUTING.md#sign-off)

### CODE REVIEW

**Comment by anoadragon453** [comment]
Can you fix the linting? Cyclomatic complexity just means the function has too many branching statements (it is indeed annoying but helps keep functions to a reasonable number of lines).

Also during testing I received a stack trace whenever sending a typing event:

```
INFO[2019-06-25T13:31:27.257914853Z] Incoming request                              req.id=hsiy5pdFCqqy req.method=PUT req.path="/_matrix/client/r0/rooms/!aaLEoOz66BAshy1K:localhost/typing/@bla:localhost"
ERRO[2019-06-25T13:31:27.258729695Z] Request panicked!
goroutine 130 [running]:
runtime/debug.Stack(0xc000392cc0, 0xc0002192b8, 0xc000392d20)
	/usr/lib/go-1.11/src/runtime/debug/stack.go:24 +0xa7
github.com/matrix-org/util.Protect.func1.1(0xc0005c2700, 0xc05ec0, 0xc000652c40)
	/home/user/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:87 +0x16f
panic(0xa6a140, 0x1070190)
	/usr/lib/go-1.11/src/runtime/panic.go:513 +0x1b9
github.com/matrix-org/dendrite/typingserver/cache.(*TypingCache).GetTypingUsersIfUpdatedAfter(0xc000251b30, 0xc00001cab9, 0x1b, 0x0, 0x8, 0xc0002196c0, 0xc000219698, 0x8)
	/home/user/code/dendrite/typingserver/cache/cache.go:66 +0x8a
github.com/matrix-org/dendrite/typingserver/cache.(*TypingCache).GetTypingUsers(0xc000251b30, 0xc00001cab9, 0x1b, 0xc000032000, 0x203000, 0x203000)
	/home/user/code/dendrite/typingserver/cache/cache.go:53 +0x48
github.com/matrix-org/dendrite/typingserver/input.(*TypingServerInputAPI).sendEvent(0xc000251b60, 0xc000280a80, 0xe, 0xc00001cab9)
	/home/user/code/dendrite/typingserver/input/input.go:60 +0x5d
github.com/matrix-org/dendrite/typingserver/input.(*TypingServerInputAPI).InputTypingEvent(0xc000251b60, 0xc06b00, 0xc00044b6b0, 0xc000280a80, 0x109bae8, 0xa03020, 0xc00018c9f0)
	/home/user/code/dendrite/typingserver/input/input.go:56 +0xff
github.com/matrix-org/dendrite/clientapi/producers.(*TypingServerProducer).Send(0xc0002824a0, 0xc06b00, 0xc00044b6b0, 0xc00001cadc, 0xe, 0xc00001cab9, 0x1b, 0xc00001ca00, 0x0, 0xc00001cab9, ...)
	/home/user/code/dendrite/clientapi/producers/typingserver.go:49 +0x1bb
github.com/matrix-org/dendrite/clientapi/routing.SendTyping(0xc0005c2800, 0xc00044b770, 0xc00001cab9, 0x1b, 0xc00001cadc, 0xe, 0xc0001fe0f0, 0xc0002824a0, 0xb36e12, 0x6, ...)
	/home/user/code/dendrite/clientapi/routing/sendtyping.go:70 +0x2bb
github.com/matrix-org/dendrite/clientapi/routing.Setup.func17(0xc0005c2800, 0xc00044b770, 0xc0001fe0f0, 0xbff7e0, 0xc00019a960, 0x0)
	/home/user/code/dendrite/clientapi/routing/routing.go:187 +0x128
github.com/matrix-org/dendrite/common.MakeAuthAPI.func1(0xc0005c2800, 0x0, 0x0, 0x0, 0x0)
	/home/user/code/dendrite/common/httpapi.go:27 +0xb6
github.com/matrix-org/util.(*jsonRequestHandlerWrapper).OnIncomingRequest(0xc000196518, 0xc0005c2800, 0xc00012f680, 0x80, 0x80, 0xb05c20)
	/home/user/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:68 +0x33
github.com/matrix-org/util.MakeJSONAPI.func1(0xc05ec0, 0xc000652c40, 0xc0005c2800)
	/home/user/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:128 +0x89
github.com/matrix-org/util.Protect.func1(0xc05ec0, 0xc000652c40, 0xc0005c2700)
	/home/user/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:92 +0x8b
net/http.HandlerFunc.ServeHTTP(0xc000283350, 0xc05ec0, 0xc000652c40, 0xc0005c2700)
	/usr/lib/go-1.11/src/net/http/server.go:1964 +0x44
github.com/matrix-org/dendrite/common.MakeExternalAPI.func1(0xc05ec0, 0xc000652c40, 0xc0005c2600)
	/home/user/code/dendrite/common/httpapi.go:40 +0x190
net/http.HandlerFunc.ServeHTTP(0xc0002e6280, 0xc05ec0, 0xc000652c40, 0xc0005c2600)
	/usr/lib/go-1.11/src/net/http/server.go:1964 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00019a780, 0xc05ec0, 0xc000652c40, 0xc0005c2600)
	/home/user/go/pkg/mod/github.com/gorilla/mux@v1.3.0/mux.go:114 +0xe0
github.com/matrix-org/dendrite/common.WrapHandlerInCORS.func1(0xc05ec0, 0xc000652c40, 0xc0005c2400)
	/home/user/code/dendrite/common/httpapi.go:114 +0x169
net/http.HandlerFunc.ServeHTTP(0xc0004a8700, 0xc05ec0, 0xc000652c40, 0xc0005c2400)
	/usr/lib/go-1.11/src/net/http/server.go:1964 +0x44
net/http.(*ServeMux).ServeHTTP(0x107dc40, 0xc05ec0, 0xc000652c40, 0xc0005c2400)
	/usr/lib/go-1.11/src/net/http/server.go:2361 +0x127
net/http.serverHandler.ServeHTTP(0xc0004b60d0, 0xc05ec0, 0xc000652c40, 0xc0005c2400)
	/usr/lib/go-1.11/src/net/http/server.go:2741 +0xab
net/http.(*conn).serve(0xc0000a6140, 0xc06a40, 0xc0001922c0)
	/usr/lib/go-1.11/src/net/http/server.go:1847 +0x646
created by net/http.(*Server).Serve
	/usr/lib/go-1.11/src/net/http/server.go:2851 +0x2f5  context=missing panic="runtime error: invalid memory address or nil pointer dereference"
INFO[2019-06-25T13:31:27.258879396Z] Responding (35 bytes)                         code=500 context=missing
```

**Comment by Cnly** [comment]
Linting, complexity and nil pointer dereference should all be fixed now.

**Comment by Cnly** [comment]
Just refined some docs. The PR is now ready for review! :)

File: syncapi/sync/request.go [hunk 1]
```diff
@@ -74,14 +76,40 @@ func getTimeout(timeoutMS string) time.Duration {
...
 	if since == "" {
 		return nil, nil
 	}
-	i, err := strconv.Atoi(since)
-	if err != nil {
-		return nil, err
+
+	posStrings := strings.Split(since, "_")
+	if len(posStrings) != 2 && len(posStrings) != 1 {
+		// A token can either be a full length one or a short (PDU-only) one
```
Comment (anoadragon453) [inline syncapi/sync/request.go]:
```suggestion
		// A token can either be full length or short (PDU-only).
```

File: syncapi/types/types.go [hunk 2]
```diff
@@ -21,12 +21,38 @@ import (
...
 
 // String implements the Stringer interface.
-func (sp StreamPosition) String() string {
-	return strconv.FormatInt(int64(sp), 10)
+func (sp SyncPosition) String() string {
+	return strconv.FormatInt(sp.PDUPosition, 10) + "_" +
+		strconv.FormatInt(sp.TypingPosition, 10)
+}
+
+// IsAfter returns whether sp refers to states newer than other
```
Comment (anoadragon453) [inline syncapi/types/types.go]: People calling this function won't know what `sp` or `other` are.

```suggestion
// IsAfter returns whether one SyncPosition refers to states newer than another SyncPosition.
```

File: syncapi/types/types.go [hunk 3]
```diff
@@ -21,12 +21,38 @@ import (
...
+func (sp SyncPosition) IsAfter(other SyncPosition) bool {
+	return sp.PDUPosition > other.PDUPosition ||
+		sp.TypingPosition > other.TypingPosition
+}
+
+// WithUpdates returns a copy of `sp` with updates represented by `other` applied.
```
Comment (anoadragon453) [inline syncapi/types/types.go]: Same, people won't know what `sp` or `other` are.

```
// WithUpdates returns a copy of the SyncPosition with updates applied from another SyncPosition.
// If the latter SyncPosition contains a field that is not 0, it is considered an update,
// and its value will replace the corresponding value in the SyncPosition WithUpdates is called on.
```

File: typingserver/input/input.go [hunk 4]
```diff
@@ -57,15 +57,21 @@ func (t *TypingServerInputAPI) InputTypingEvent(
...
 	}
 	ote := &api.OutputTypingEvent{
-		Event:       *ev,
-		TypingUsers: userIDs,
+		Event: *ev,
+	}
+
+	if ev.Typing {
+		expireTime := ite.OriginServerTS.Time().Add(
+			time.Duration(ite.Timeout) * time.Millisecond,
```
Comment (anoadragon453) [inline typingserver/input/input.go]: I wonder: should the timeout be something that is attached to every typing event, or should it be a global constant?

Comment (anoadragon453) [reply typingserver/input/input.go]: Certainly looks quite good. I just have a few questions and suggestions.

**Comment by anoadragon453** [comment]
Also, please merge master into this branch.

(The fact testfile is so easy to conflict on makes me a little worried).

File: typingserver/input/input.go (hunk 4 above)
Comment (Cnly) [reply typingserver/input/input.go]: The client can either include or omit the `timeout` field when sending a typing event. The server only applies a default value when the field is omitted.

File: syncapi/sync/notifier.go [hunk 5]
```diff
@@ -98,11 +103,13 @@ func (n *Notifier) OnNewEvent(ev *gomatrixserverlib.Event, userID string, pos ty
...
-		}
-	} else if len(userID) > 0 {
-		n.wakeupUser(userID, pos)
+		n.wakeupUsers(usersToNotify, latestPos)
+	} else if roomID != "" {
+		n.wakeupUsers(n.joinedUsers(roomID), latestPos)
+	} else if len(userIDs) > 0 {
+		n.wakeupUsers(userIDs, latestPos)
+	} else {
+		log.Warn("WARNING: Notifier.OnNewEvent called but caller supplied no user to wake up")
```
[the remark this answers is not in the document]
Comment (Cnly) [reply syncapi/sync/notifier.go]: I think `posUpdate` may be more suitable for tracing the cause.

File: typingserver/input/input.go (hunk 4 above)
Comment (anoadragon453) [reply typingserver/input/input.go]: Ah, yes I see. Missed that bit of the spec.

File: syncapi/storage/syncserver.go [hunk 6]
```diff
@@ -257,52 +268,151 @@ func (d *SyncServerDatabase) IncrementalSync(
...
+		fromPos, toPos, joinedRoomIDs, res,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// getResponseWithPDUsForCompleteSync creates a response and add all PDUs needed
```
Comment (anoadragon453) [inline syncapi/storage/syncserver.go]:
```suggestion
// getResponseWithPDUsForCompleteSync creates a response and adds all PDUs needed
```

File: syncapi/storage/syncserver.go [hunk 7]
```diff
@@ -385,10 +517,30 @@ func (d *SyncServerDatabase) RetireInviteEvent(
 	return err
 }
 
-func (d *SyncServerDatabase) addInvitesToResponse(
+func (d *SyncServerDatasource) SetTypingTimeoutCallback(fn cache.TimeoutCallbackFn) {
+	d.typingCache.SetTimeoutCallback(fn)
+}
+
+// AddTypingUser adds a typing user to the typing cache.
+// Returns the latest sync position for typing notifications after update.
```
Comment (anoadragon453) [inline syncapi/storage/syncserver.go]:
```suggestion
// Returns the newly calculated sync position for typing notifications.
```

File: syncapi/storage/syncserver.go [hunk 8]
```diff
@@ -385,10 +517,30 @@ func (d *SyncServerDatabase) RetireInviteEvent(
...
+func (d *SyncServerDatasource) AddTypingUser(
+	userID, roomID string, expireTime *time.Time,
+) int64 {
+	return d.typingCache.AddTypingUser(userID, roomID, expireTime)
+}
+
+// RemoveTypingUser removes a typing user from the typing cache.
+// Returns the latest sync position for typing notifications after update.
```
Comment (anoadragon453) [inline syncapi/storage/syncserver.go]:
```suggestion
// Returns the newly calculated sync position for typing notifications.
```

File: syncapi/sync/notifier.go [hunk 9]
```diff
@@ -58,20 +58,30 @@ func NewNotifier(pos types.StreamPosition) *Notifier {
 
 // OnNewEvent is called when a new event is received from the room server. Must only be
 // called from a single goroutine, to avoid races between updates which could set the
-// current position in the stream incorrectly.
-// Can be called either with a *gomatrixserverlib.Event, or with an user ID
-func (n *Notifier) OnNewEvent(ev *gomatrixserverlib.Event, userID string, pos types.StreamPosition) {
+// current sync position incorrectly.
+// Chooses which user sync streams to update by a provided *gomatrixserverlib.Event
+// (based on the users in the event's room),
+// a roomID directly, or a list of user IDs, whichever specified first in the parameters.
```
Comment (anoadragon453) [inline syncapi/sync/notifier.go]:
```suggestion
// a roomID directly, or a list of user IDs, prioritised by parameter ordering.
```

File: syncapi/sync/notifier.go [hunk 10]
```diff
@@ -58,20 +58,30 @@ func NewNotifier(pos types.StreamPosition) *Notifier {
...
+// posUpdate contains the latest position(s) for one or more types of events.
+// If a position in posUpdate is 0, it means no updates available of that type.
```
Comment (anoadragon453) [inline syncapi/sync/notifier.go]:
```suggestion
// If a position in posUpdate is 0, it means no updates are available of that type.
```

File: syncapi/sync/notifier.go [hunk 11]
```diff
@@ -58,20 +58,30 @@ func NewNotifier(pos types.StreamPosition) *Notifier {
...
+// Typically a consumer supplies a posUpdate with the sync position for the
```
Comment (anoadragon453) [inline syncapi/sync/notifier.go]:
```suggestion
// Typically a consumer supplies a posUpdate with the latest sync position for the
```

File: syncapi/sync/notifier.go [hunk 12]
```diff
@@ -58,20 +58,30 @@ func NewNotifier(pos types.StreamPosition) *Notifier {
...
+// event type it handles set to the latest, leaving other fields as 0.
```
Comment (anoadragon453) [inline syncapi/sync/notifier.go]:
```suggestion
// event type it handles, leaving other fields as 0.
```

File: typingserver/cache/cache.go [hunk 13]
```diff
@@ -22,25 +22,66 @@ const defaultTypingTimeout = 10 * time.Second
 // userSet is a map of user IDs to a timer, timer fires at expiry.
 type userSet map[string]*time.Timer
 
+// TimeoutCallbackFn is a function called right after removal of a user
```
Comment (anoadragon453) [inline typingserver/cache/cache.go]:
```suggestion
// TimeoutCallbackFn is a function called right after the removal of a user
```

**Pull request merged**